iscservice/isc.c

```diff
@@ -680,6 +689,250 @@ M0_INTERNAL int m0_isc_comp_state_probe(const struct m0_fid *fid)
+		bv->ov_vec.v_count[i] >>= shift;
+		bv->ov_buf[i] = m0_stob_addr_pack(bv->ov_buf[i], shift);
+	}
+}
+
+static void bufvec_open(struct m0_bufvec *bv, uint32_t shift)
+{
+	uint32_t i;
+
+	for (i = 0; i < bv->ov_vec.v_nr; i++) {
+		bv->ov_vec.v_count[i] <<= shift;
+		bv->ov_buf[i] = m0_stob_addr_open(bv->ov_buf[i], shift);
+	}
+}
+
+static int bufvec_alloc_init(struct m0_bufvec *bv, struct m0_io_indexvec *iiv,
+			     uint32_t shift)
+{
+	int   rc;
+	int   i;
+	char *p;
+
+	p = m0_alloc_aligned(m0_io_count(iiv), shift);
+	if (p == NULL)
+		return M0_ERR_INFO(-ENOMEM, "failed to allocate buf");
+
+	rc = m0_bufvec_empty_alloc(bv, iiv->ci_nr);
+	if (rc != 0) {
+		m0_free(p);
+		return M0_ERR_INFO(rc, "failed to allocate bufvec");
+	}
+
+	for (i = 0; i < iiv->ci_nr; i++) {
+		bv->ov_buf[i] = p;
+		bv->ov_vec.v_count[i] = iiv->ci_iosegs[i].ci_count;
+		p += iiv->ci_iosegs[i].ci_count;
+	}
+
+	return 0;
+}
+
+M0_INTERNAL int m0_isc_io_launch(struct m0_stob_io *stio,
+				 struct m0_fid *cob,
+				 struct m0_io_indexvec *iv,
+				 struct m0_fom *fom)
+{
+	int                rc;
+	uint32_t           shift = 0;
+	struct m0_stob_id  stob_id;
+	struct m0_stob    *stob = NULL;
+
+	if (iv->ci_nr == 0) {
+		rc = M0_ERR_INFO(-EINVAL, "no io segments given");
+		goto err;
+	}
+
+	m0_stob_io_init(stio);
+	stio->si_opcode = SIO_READ;
+
+	m0_fid_convert_cob2stob(cob, &stob_id);
+	rc = m0_storage_dev_stob_find(m0_cs_storage_devs_get(),
+				      &stob_id, &stob);
+	if (rc != 0) {
+		M0_ERR_INFO(rc, "failed to find stob by cob="FID_F, FID_P(cob));
+		goto err;
+	}
+
+	shift = m0_stob_block_shift(stob);
+	stio->si_obj = stob; /* for m0_isc_io_fini() in case of err */
+
+	rc = m0_indexvec_wire2mem(iv, iv->ci_nr, shift, &stio->si_stob);
+	if (rc != 0) {
+		M0_ERR_INFO(rc, "failed to make cob ivec");
+		goto err;
+	}
+
+	rc = bufvec_alloc_init(&stio->si_user, iv, shift);
+	if (rc != 0) {
+		M0_ERR_INFO(rc, "failed to allocate bufvec");
+		goto err;
+	}
+
+	bufvec_pack(&stio->si_user, shift);
+
+	rc = m0_stob_io_private_setup(stio, stob);
+	if (rc != 0) {
+		M0_ERR_INFO(rc, "failed to setup adio for stob="FID_F,
+			    FID_P(&stob->so_id.si_fid));
+		goto err;
+	}
+
+	/* make sure the fom is waked up on I/O completion */
+	m0_mutex_lock(&stio->si_mutex);
+	m0_fom_wait_on(fom, &stio->si_wait, &fom->fo_cb);
+	m0_mutex_unlock(&stio->si_mutex);
+
+	rc = m0_stob_io_prepare_and_launch(stio, stob, NULL, NULL);
+	if (rc != 0) {
+		m0_mutex_lock(&stio->si_mutex);
+		m0_fom_callback_cancel(&fom->fo_cb);
+		m0_mutex_unlock(&stio->si_mutex);
+		M0_ERR_INFO(rc, "failed to launch io for stob="FID_F,
+			    FID_P(&stob->so_id.si_fid));
+	}
+ err:
+	if (rc != 0 && stob != NULL) {
+		bufvec_open(&stio->si_user, shift);
+		m0_isc_io_fini(stio);
+	}
+
+	return rc;
+}
+
+M0_INTERNAL int64_t m0_isc_io_res(struct m0_stob_io *stio, char **buf)
+{
+	uint32_t shift = m0_stob_block_shift(stio->si_obj);
+
+	bufvec_open(&stio->si_user, shift);
+	*buf = stio->si_user.ov_buf[0];
```

Comment:
By this, we are only using the first vector in this result.
We are assuming the first vector holds all the results.

Can we use the "stio->si_user" and use a m0_bufvec_cursor to iterate over it?

Reply:
We allocated only a single buffer for the I/O (see `bufvec_alloc_init()`), so only one is possible here, really.